virtio-devices/src/block.rs

```diff
@@ -541,13 +554,34 @@ impl EpollHelperHandler for BlockEpollHandler {
                     )));
                 }
             }
+            PROXY_KILL_EVENT => {
+                if self.pending_kill == ProxyKillEventState::Running {
+                    self.pending_kill = ProxyKillEventState::Pending;
+                }
+            }
             _ => {
                 return Err(EpollHelperError::HandleEvent(anyhow!(
                     "Unexpected event: {}",
                     ev_type
                 )));
             }
         }
+
+        // Pending kill and no more inflight requests -> kill for real
+        if self.pending_kill == ProxyKillEventState::Pending {
+            if self.inflight_requests.is_empty() {
+                self.kill_evt.write(1).map_err(|e| {
+                    EpollHelperError::HandleEvent(anyhow!("Failed to write kill event: {:?}", e))
+                })?;
+                self.pending_kill = ProxyKillEventState::Sent;
```

Comment:
Can you please explain more why a tri-state variable is needed?

Wouldn't a boolean is enough? Say, `pending_kill` starts with `False`, and is set to `True` when such proxy event is  triggered. Once all inflight IOs are completed, we write the real `kill_evt` and set `pending_kill` back to `False` again.

Reply:
This is to avoid writing to `kill_evt` a second time - remember `EventFd` semantics differ compared to ordinary file descriptors.